pkg/mimirtool/commands/analyse_prometheus.go

```diff
@@ -34,173 +38,197 @@ type PrometheusAnalyzeCommand struct {
+	if err := parseMetricFileIfExist(cmd.rulerMetricsFile, rulerMetrics); err != nil {
+		return nil, err
 	}
+	metricsUsed = append(metricsUsed, rulerMetrics.MetricsUsed...)
 
-	if !hasGrafanaMetrics && !hasRulerMetrics {
-		return errors.New("No Grafana or Ruler metrics files")
+	if len(metricsUsed) == 0 {
+		return nil, errors.New("no Grafana or Ruler metrics files")
 	}
 
+	return metricsUsed, nil
+}
+
+func (cmd *PrometheusAnalyzeCommand) init() (v1.API, error) {
 	rt := api.DefaultRoundTripper
 	if cmd.username != "" {
 		rt = config.NewBasicAuthRoundTripper(cmd.username, config.Secret(cmd.password), "", api.DefaultRoundTripper)
 	}
-	promClient, err := api.NewClient(api.Config{
+
+	client, err := api.NewClient(api.Config{
 		Address:      cmd.address,
 		RoundTripper: rt,
 	})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	v1api := v1.NewAPI(promClient)
+	return v1.NewAPI(client), nil
+}
 
+func (cmd *PrometheusAnalyzeCommand) queryMetricNames(apiWithBackoff withBackoffFn) (model.LabelValues, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), cmd.readTimeout)
 	defer cancel()
-	metricNames, _, err := v1api.LabelValues(ctx, labels.MetricName, nil, time.Now().Add(-10*time.Minute), time.Now())
+
+	var metricNames model.LabelValues
+	err := apiWithBackoff(ctx, func(ctx context.Context, api v1.API) error {
+		var err error
+		metricNames, _, err = api.LabelValues(ctx, labels.MetricName, nil, time.Now().Add(-10*time.Minute), time.Now())
+		return err
+	})
 	if err != nil {
-		return errors.Wrap(err, "error querying for metric names")
+		return nil, errors.Wrap(err, "error querying for metric names")
 	}
-	log.Infof("Found %d metric names\n", len(metricNames))
 
-	inUseMetrics := map[string]struct {
-		totalCount int
-		jobCount   map[string]int
-	}{}
-	inUseCardinality := 0
-	var errorMetrics []string
-	for _, metric := range metricsUsed {
-		ctx, cancel := context.WithTimeout(context.Background(), cmd.readTimeout)
-		defer cancel()
-
-		query := "count by (job) (" + metric + ")"
-		result, _, err := v1api.Query(ctx, query, time.Now())
-		if err != nil {
-			errStr := fmt.Sprintf("skipped %s analysis because failed to run query %v: %s", metric, query, err.Error())
-			log.Warnln(errStr)
-			errorMetrics = append(errorMetrics, errStr)
-			continue
-		}
-
-		vec := result.(model.Vector)
-		if len(vec) == 0 {
-			counts := inUseMetrics[metric]
-			counts.totalCount += 0
-			inUseMetrics[metric] = counts
-			log.Debugln("in use", metric, 0)
-
-			continue
-		}
-
-		for _, sample := range vec {
-			counts := inUseMetrics[metric]
-			if counts.jobCount == nil {
-				counts.jobCount = make(map[string]int)
-			}
+	log.Infof("Found %d metric names\n", len(metricNames))
+	return metricNames, nil
+}
 
-			counts.totalCount += int(sample.Value)
-			counts.jobCount[string(sample.Metric["job"])] += int(sample.Value)
-			inUseMetrics[metric] = counts
+type analyse struct {
```

Comment:
Can we use `analysisResult` or another noun instead of `analyze`?

Reply:
If so, can you update the variables like `inUseMetricsAnalyze` to `inUseMetricsAnalysisResult` for example?